src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -5162,6 +5200,27 @@ internal EventDispatcher(EventDispatcher? next, bool[]? eventEnabled, EventListe
         internal readonly EventListener m_Listener;   // The dispatcher this entry is for
         internal bool[]? m_EventEnabled;              // For every event in a the eventSource, is it enabled?
 
+        /// Filtering logic for self-describing EventSources
+        /// Events coming from self-describing EventSources have EventIDs of -1, so they need to be filtered separately
+        /// via their event level and keywords.
+        internal EventLevel m_Level;         // Level at which this EventListener enabled this EventSource
+        internal EventKeywords m_Keywords;   // Keyword with which this EventListener enabled this EventSource
+
+        internal void UpdateLevelAndKeyword(EventLevel level, EventKeywords keywords)
+        {
+            m_Level = level;
+            m_Keywords = keywords;
+        }
+
+        internal bool IsEventEnabled(EventLevel level, EventKeywords keywords)
+        {
+            if (level == EventLevel.LogAlways || keywords == 0 || m_Keywords == 0)
```

Comment:
This logic still has some trouble. One example would be m_level =Informational, m_keywords=0x0, level = verbose, keywords=0x0. Expected result: false (listener level is too low), Actual result: true.

You probably should be calling EventSource.IsEnabledCommon() except reviewing that code it also appears to have bugs [here](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L2355) and [here](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L2359) : (. EventPipe native code has a [correct implementation](https://github.com/dotnet/runtime/blob/master/src/coreclr/src/vm/eventpipeconfiguration.cpp#L328) though the special case in the levelEnabled check for LogAlways is unnecessary complexity. LogAlways is the lowest value so sessionLevel >= LogAlways is always true.

The fact that we've got all these all these filtering bugs lurking probably means we are in need of EventListener level/keyword filtering tests for both the self-describing case and the manifested events.

I think you could fix IsEnabledCommon() to look like this:
````
        private static bool IsEnabledCommon(bool enabled, EventLevel currentLevel, EventKeywords currentMatchAnyKeyword,
                                                          EventLevel eventLevel, EventKeywords eventKeywords, EventChannel eventChannel)
        {
            if (!enabled)
                return false;

            // does is pass the level test?
            if (currentLevel < eventLevel)
                return false;

            // if yes, does it pass the keywords test?
            if (eventKeywords != 0)
            {
#if FEATURE_MANAGED_ETW_CHANNELS
                // is there a channel with keywords that match currentMatchAnyKeyword?
                if (eventChannel != EventChannel.None && this.m_channelData != null && this.m_channelData.Length > (int)eventChannel)
                {
                    EventKeywords channel_keywords = unchecked((EventKeywords)(m_channelData[(int)eventChannel] | (ulong)eventKeywords));
                    if (channel_keywords != 0 && (channel_keywords & currentMatchAnyKeyword) == 0)
                        return false;
                }
                else
#endif
                {
                    if ((unchecked((ulong)eventKeywords & (ulong)currentMatchAnyKeyword)) == 0)
                        return false;
                }
            }
            return true;
        }
````
Once you had that you could invoke it with Channel.None to get the right filtering behavior.

Reply:
I also notice the code in WriteStringToAllListeners() has a suspicious loop over m_EventEnabled[] that tries to infer if the listener is enabled or not. I could imagine that code may also not be doing the right thing for self-describing events? It would worth testing as part of the filtering tests.